tests/test_recursive.py

```diff
@@ -0,0 +1,253 @@
+        assert config.mode == "mode"
+        assert config.sub.a == 42
+        assert config.sub.b == "abc"
+        assert config.flag is True
+
+    def test_set_sub(self):
+        config = parse(self.Config, ["mode", "--sub-a", "1"])
+        assert config.mode == "mode"
+        assert config.sub.a == 1
+        assert config.sub.b == "abc"
+        assert config.flag is False
+
+
+class TestParseSubConfigWithRequiredField:
+    @dataclass
+    class Config:
+        mode: str = field(metadata=dict(positional=True))
+        sub: SubConfigWithRequiredField
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    def test_parse_required(self, capsys):
+        with raises(SystemExit):
+            parse(self.Config, ["mode"])
+        captured = capsys.readouterr()
+        assert "the following arguments are required: --sub-a" in captured.err
+
+    def test_flag(self):
+        config = parse(self.Config, ["mode", "--sub-a", "1", "--flag"])
+        assert config.mode == "mode"
+        assert config.sub.a == 1
+        assert config.sub.b == "abc"
+        assert config.flag is True
+
+    def test_set_sub(self):
+        config = parse(self.Config, ["mode", "--sub-a", "1", "--sub-b", "def"])
+        assert config.mode == "mode"
+        assert config.sub.a == 1
+        assert config.sub.b == "def"
+        assert config.flag is False
+
+
+class TestParseSubConfigWithPositional:
+    @dataclass
+    class Config:
+        mode: str = field(metadata=dict(positional=True))
+        sub: SubConfigWithPositional = field(default_factory=SubConfigWithPositional)
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    def test_help(self, capsys):
+        with raises(SystemExit):
+            parse(self.Config, ["--help"], prog="prog")  # type: ignore
+        captured = capsys.readouterr()
+        assert "mode [sub_a]" in captured.out
+
+    def test_parse(self):
+        config = parse(self.Config, ["mode"])
+        assert config.mode == "mode"
+        assert config.sub.a == 42
+        assert config.sub.b == "abc"
+        assert config.flag is False
+
+    def test_flag(self):
+        config = parse(self.Config, ["mode", "1", "--flag"])
+        assert config.mode == "mode"
+        assert config.sub.a == 1
+        assert config.sub.b == "abc"
+        assert config.flag is True
+
+    def test_set_sub(self):
+        config = parse(
+            self.Config,
+            ["mode", "1", "--sub-b", "def"],
+        )
+        assert config.mode == "mode"
+        assert config.sub.a == 1
+        assert config.sub.b == "def"
+        assert config.flag is False
+
+
+class TestParsePositionalSubConfig:
+    @dataclass
+    class Config:
+        mode: str = field(metadata=dict(positional=True))
+        sub: SubConfig = field(default_factory=SubConfig, metadata=dict(positional=True))
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    def test_parse(self):
```

Comment:
is it possible to add a capsys check here?

Reply:
Sure - but there will be no output. This raises an exception, which differs from the `SystemExit` raised by argparse after it prints a helpful message.

Reply:
Ah okay that's less useful indeed